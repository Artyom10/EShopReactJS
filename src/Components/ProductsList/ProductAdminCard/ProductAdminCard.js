import React from 'react';

const stylesForAdminCard = {
    cardStyle: {
        maxWidth: '540px'
    },

    colInContainer: {
        flex: '1 0 50%',
        paddingTop: '12px',
        paddingBottom: '12px'
    }
}

function ProductAdminCard(props) {
    const {product} = props; 
  return (
    <div className="col d-md-flex" style={stylesForAdminCard.colInContainer}>
    <div className="card mb-3" style={stylesForAdminCard.cardStyle}>
      <div className="row no-gutters justify-content-center">
        <div className="col-md-6">
          <img src={product.urlPhoto} className="card-img" alt={product.type} />
        </div>
        <div className="col-md-8">
          <div className="card-body">
            <h5 className="card-title">Main information</h5>
            <ul className="list-group">
              <form>
                <div className="form-group">
                  <label for="price">Price</label>
                  <input className="form-control" type="text" id="price" value={product.price} />
                  <label for="producer">Producer</label>
                  <input type="text" className="form-control" id="producer" value={product.producer} />
                  <label for="type">Type</label>
                  <input type="text" className="form-control" id="type" value={product.type} />
                  <label for="sizes">Sizes</label>
                  <input className="form-control" type="text" id="sizes" value={product.sizes} />
                  <label for="description">Description</label>
                  <textarea className="form-control" id="description" rows="3">{product.description}</textarea>
                  <label for="tags">Tags</label>
                  <input type="text" className="form-control" id="tags" value={product.tags} />
                </div>
            </form>
              <li className="list-group-item">
              <div className="row justify-content-center">
                <div className="col">
                  <button className="btn general-button edit-button">Edit product</button>
                </div>
                <div className="col">
                  <button className="btn general-button delete-button">Delete product</button>
                </div>
              </div>
            </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  );
}

export default ProductAdminCard;
