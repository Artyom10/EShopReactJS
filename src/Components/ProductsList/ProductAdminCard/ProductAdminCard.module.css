.colInContainer {
    flex: 1 0 50%;
    padding-top: 12px;
    padding-bottom: 12px;
}

.cardStyle {
    max-width: 540px
}

.editButton {
    background: #2796ff;
  }
  
  .deleteButton {
    background: #ed3330;
  }

  .deleteButton:hover,
  .editButton:hover,
  .addButton:hover:hover {
    background: #434343;
    letter-spacing: 1px;
    -webkit-box-shadow: 0px 5px 40px -10px rgba(0, 0, 0, 0.57);
    -moz-box-shadow: 0px 5px 40px -10px rgba(0, 0, 0, 0.57);
    box-shadow: 5px 40px -10px rgba(0, 0, 0, 0.57);
    transition: all 0.4s ease 0s;
  }

  .insideCard{
    height: 400px;
    background-size: cover;
  }

