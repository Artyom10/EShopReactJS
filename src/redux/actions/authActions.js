export const signIn = (credentials) => {
    debugger;
    return (dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signInWithEmailAndPassword(
            credentials.email,
            credentials.password
        ).then(() => {
            dispatch({type: 'LOGIN_SUCCESS'});
        }).catch((err) => {
            dispatch({type: 'LOGIN_ERROR', err})
        });
    }
};

export const signOut = () => {
    return(dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signOut().then(() => {
            dispatch({type: 'SIGNOUT_SUCCESS'})
        });
    } 
}

export const signUp = (newUser) => {
    return(dispatch, getState, {getFirebase, getFirestore}) => {
        debugger;
        const firebase = getFirebase();
        const firestore = getFirestore();

        firebase.auth().createUserWithEmailAndPassword(
            newUser.email,
            newUser.password,
        ).then((response) => {
            return firestore.collection('users').doc(response.user.uid).set({
                firstName: newUser.firstName,
                secondName: newUser.secondName,
                request: false, //newUser.request,
                bags: [], //newUser.bags,
                userPhoto: 'https://icons-for-free.com/iconfiles/png/512/business+costume+male+man+office+user+icon-1320196264882354682.png',
                valuedProducts: [],//newUser.valuedProducts,
                isAdmin: false,
                isWho: 'user',
                email: newUser.email,
            })
        }).then(() => {
            dispatch({type: 'SIGNUP_SUCCESS'})
        }).catch(err => {
            dispatch({type: 'SIGNUP_ERROR', err})
        })
    }
}