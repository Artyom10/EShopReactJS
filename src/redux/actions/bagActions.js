/*export const removeFromBag = (removeTargetBag) => {
    return (dispatch, getState) => {
 dispatch({ type: 'REMOVE_FROM_BAG', removeTargetBag,})
    }
}*/

export const addToBag = (targetAddToBag) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     //const profile = getState()
     const buyerId = getState().firebase.auth.uid
     firestore.collection('users').doc(buyerId).update({
            bags: [targetAddToBag]
     })
     .then(() => {
         dispatch({type: 'ADD_TO_BAG', targetAddToBag})
     }).catch((err) => {
         dispatch({type: 'ADD_TO_BAG_ERROR', err})
     })
    }
}


 //dispatch({type: 'ADD_TO_BAG', targetAddToBag})