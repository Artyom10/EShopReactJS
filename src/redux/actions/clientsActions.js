import API from '../../services/axios';

export const deleteUser = (uid) => {
    return(dispatch,getState,{getfirebase, getFirestore}) => {
        API.post('clients', uid);
    }

}