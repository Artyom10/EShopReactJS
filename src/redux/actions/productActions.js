export const addNewProduct = (newProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();

    const infoPhoto = newProduct.photo;
    newProduct.photo = null;

    firestore
      .collection("products")
      .add({
        ...newProduct,
        isBought: false,
      })
      .then((response) => {
        infoPhoto &&
        storage
          .ref(`photos/${response.id}`)
          .put(infoPhoto)
          .on(
            'state_changed',
            snapshot => { },
            error => { console.log(error) },
            () => {
              storage
                .ref('photos')
                .child(response.id)
                .getDownloadURL()
                .then(url => {
                  newProduct.photo = url
                  return firestore
                    .collection('products')
                    .doc(response.id)
                    .update({photo: newProduct.photo})
                })
            }
          )
      })
      .then(() => {
        dispatch({ type: "ADD_PRODUCT", newProduct });
      })
      .catch((err) => {
        dispatch({ type: "ADD_PRODUCT_ERROR", err });
      });
  };
};

export const removeProduct = (targetDeleteProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    console.log(getState());
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    firestore
      .collection("products")
      .doc(targetDeleteProduct)
      .get()
      .then((doc) => {
        firestore
          .collection('products')
          .doc(targetDeleteProduct)
          .delete()
          .then(() => {
            doc.data().photo && 
            storage
              .ref(`photos/${targetDeleteProduct}`)
              .delete()
          })
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT", targetDeleteProduct });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });

      const buyerId = getState().firebase.auth.uid;
      const firebaseBuyer = getState().firebase;
      let findIndex = null;

      let ratedProducts = firebaseBuyer.profile.valuedProducts;
      ratedProducts.forEach((product,index) => {
        if(product.targetProductRating === targetDeleteProduct){
                findIndex = index;       
        }
      });
      ratedProducts.splice(findIndex,1);

      let findBagsIndex = null;

      let bagsProducts = firebaseBuyer.profile.bags;
      bagsProducts.forEach((product,index) => {
        if(product === targetDeleteProduct){
          findBagsIndex = index
        }
      });
      bagsProducts.splice(findBagsIndex,1);
  
      firestore.collection('users').doc(buyerId).update({
       valuedProducts:   ratedProducts && [...ratedProducts],
       bags: bagsProducts && [...bagsProducts],
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT" });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });


  };
};

export const editProduct = (targetEditProduct, changes) => {
  debugger;
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    typeof changes.photo !== 'string' &&  changes.photo
    ?
    storage
      .ref('photos')
      .child(targetEditProduct)
      .put(changes.photo)
      .on(
        'state_changed',
        snapshot => { },
        error => { console.log(error) },
        () => {
          storage
           .ref('photos')
           .child(targetEditProduct)
           .getDownloadURL()
           .then(url => {
             changes.photo = url
             return firestore
               .collection('products')
               .doc(targetEditProduct)
               .update({
                 ...changes
               })
           })
           .then(() => {
            dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
          })
          .catch((err) => {
            dispatch({ type: "EDIT_PRODUCT_ERROR", err });
          });
        }
      )
      :
      firestore
        .collection('products')
        .doc(targetEditProduct)
        .update({
          ...changes
        })
        .then(() => {
          dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
        })
        .catch((err) => {
          dispatch({ type: "EDIT_PRODUCT_ERROR", err });
        });
  };
};

export const buyProduct = (targetProductBuy) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    firestore
      .collection("products")
      .doc(targetProductBuy)
      .update({
        isBought: true,
      })
      .then(() => {
        dispatch({ type: "UPDATE_BOUGHT_PRODUCT_STATUS", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "AUPDATE_BOUGHT_PRODUCT_STATUS_ERROR", err });
      });

    firestore
      .collection("users")
      .doc(buyerId)
      .update({
        bags: firebaseBuyer.profile.bags
          ? [...firebaseBuyer.profile.bags, targetProductBuy]
          : [targetProductBuy],
      })
      .then(() => {
        dispatch({ type: "ADD_TO_BOUGHT", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "ADD_TO_BOUGHT_ERROR", err });
      });
  };
};

export const setRating = (targetProductRating, value, certainProduct) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
  
    if(firebaseBuyer.profile.valuedProducts.some((everyData) => everyData.targetProductRating === targetProductRating)){
    console.log('You have already rated this product, please delete your rate and try again');
    }
    else{
          console.log('This product does not rate yet, we need update product and value')
          firestore.collection('users').doc(buyerId).update({
            valuedProducts: [...firebaseBuyer.profile.valuedProducts, {targetProductRating, value,}],
        })
        .then(() => {
          dispatch({ type: "RATE_PRODUCT", targetProductRating });
        })
        .catch((err) => {
          dispatch({ type: "RATE_PRODUCT_ERROR", err });
        });
    }
  }
}

export const deleteRating = (targetProductDeleteRating) => {
  debugger;
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let ratedProducts = firebaseBuyer.profile.valuedProducts;
    ratedProducts.forEach((product,index) => {
      if(product.targetProductRating === targetProductDeleteRating){
              findIndex = index;       
      }
    });
    ratedProducts.splice(findIndex,1);

    firestore.collection('users').doc(buyerId).update({
     valuedProducts:   ratedProducts && [...ratedProducts]
    })
    .then(() => {
      dispatch({ type: "DELETE_RATE_PRODUCT", targetProductDeleteRating });
    })
    .catch((err) => {
      dispatch({ type: "DELETE_RATE_PRODUCT_ERROR", err });
    });
  }
}


export const deleteBooked = (productId) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    debugger;
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let bagsProducts = firebaseBuyer.profile.bags;
    bagsProducts.forEach((product,index) => {
      if(product === productId){
              findIndex = index;       
      }
    });
    bagsProducts.splice(findIndex,1);

    firestore.collection('products').doc(productId).update({
      isBought: false,
    })
    .catch((err) => {
      dispatch({type: "DELETE_BAG_PRODUCT_ERROR", err});
    })
    
    firestore.collection('users').doc(buyerId).update({
      bags:   bagsProducts && [...bagsProducts]
     })
     .then(() => {
       dispatch({ type: "DELETE_BAG_PRODUCT" });
     })
     .catch((err) => {
       dispatch({ type: "DELETE_BAG_PRODUCT_ERROR", err });
     });
   }
}