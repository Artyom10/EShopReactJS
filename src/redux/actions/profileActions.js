export const updateProfile = (firstName, secondName) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     const profile = getState().firebase.auth.uid
     firestore.collection('users').doc(profile).update({
            firstName,
            secondName,
     })
     .then(() => {
         dispatch({type: 'UPDATE_PROFILE'})
     }).catch((err) => {
         dispatch({type: 'UPDATE_PROFILE_ERROR', err})
     })
    }
}

export const removeRequest = (user) => {
    debugger;
    return (dispatch, getState, {getFirebase, getFirestore}) => {
        const firestore = getFirestore();
        const profile = getState().firebase.auth.uid
        firestore.collection('users').doc(profile).update({
            request: true,
        })
        .then(() => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST'})
        }).catch((err) => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST_ERROR', err})
        })
    }
}