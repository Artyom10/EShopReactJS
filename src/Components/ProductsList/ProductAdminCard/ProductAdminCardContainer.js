import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { firestoreConnect } from 'react-redux-firebase';
import ProductAdminCard from './ProductAdminCard';
import { editProduct, removeProduct } from '../../../redux/actions/productActions';


const mapStateToProps = (state) => {
    return {
        products: state.firestore.ordered.products || state.productPages.products,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
      removeProduct: (targetDeleteProduct) => dispatch(removeProduct(targetDeleteProduct)),
      editProduct: (targetEditProduct, newData) => dispatch(editProduct(targetEditProduct, newData)),
    
    }
  }

export default connect(mapStateToProps,mapDispatchToProps)(ProductAdminCard);