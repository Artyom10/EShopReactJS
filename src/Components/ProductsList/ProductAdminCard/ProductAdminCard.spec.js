import { shallow } from 'enzyme';
import React from 'react';
import ProductAdminCard from './ProductAdminCard';
import { editProduct, removeProduct } from '../../../redux/actions/productActions';



const props = {
   editProduct,
   removeProduct,
   products: [
       {id: 1, producer: 'Fred Perry', price: '270', type: 'Olympic',  photo: null,
       description: `As far as I'm concerd it's your favorite clothes`, isBought: false, size: 'M', rated: [],},
       {id: 2, producer: 'Fashion', price: '5000', type: 'Jacket',  photo: null,
       description: `Always here`, isBought: false, size: 'L', rated: [],},
       {id: 3, producer: 'Milk', price: '10', type: 'T-Shirt',  photo: null,
       description: `That's it`, isBought: true, size: 'XS', rated: [],},
       
   ]
}

const setUp = (props) => shallow(<ProductAdminCard {...props} />);

describe('Profile component', () => {
    it('should render Profile component', () => {
       const component = setUp(props);
       expect(component.debug()).toMatchSnapshot();
    });
});