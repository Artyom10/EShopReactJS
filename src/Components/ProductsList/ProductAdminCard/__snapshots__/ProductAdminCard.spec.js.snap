// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component should render Profile component 1`] = `
"<div className=\\"col d-md-flex colInContainer\\">
  <div className=\\"card mb-3 cardStyle\\">
    <div className=\\"row no-gutters justify-content-center\\">
      <div className=\\"col-md-6\\">
        <img src={{...}} className=\\"card-img insideCard\\" alt=\\"Olympic\\" />
      </div>
      <div className=\\"col-md-8\\">
        <div className=\\"card-body\\">
          <h5 className=\\"card-title\\">
            Main information
          </h5>
          <ul className=\\"list-group\\">
            <form>
              <div className=\\"form-group\\" />
            </form>
            <li className=\\"list-group-item\\">
              <div className=\\"row justify-content-center\\">
                <div className=\\"col\\" />
                <div className=\\"col\\">
                  <button className=\\"btn generalButton deleteButton\\" onClick={[Function: onClick]}>
                    Delete product
                  </button>
                </div>
                <div className=\\"col\\">
                  <button className=\\"btn generalButton allowButton\\" onClick={[Function: onClick]}>
                    Allow edit
                  </button>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>


<div className=\\"col d-md-flex colInContainer\\">
  <div className=\\"card mb-3 cardStyle\\">
    <div className=\\"row no-gutters justify-content-center\\">
      <div className=\\"col-md-6\\">
        <img src={{...}} className=\\"card-img insideCard\\" alt=\\"Jacket\\" />
      </div>
      <div className=\\"col-md-8\\">
        <div className=\\"card-body\\">
          <h5 className=\\"card-title\\">
            Main information
          </h5>
          <ul className=\\"list-group\\">
            <form>
              <div className=\\"form-group\\" />
            </form>
            <li className=\\"list-group-item\\">
              <div className=\\"row justify-content-center\\">
                <div className=\\"col\\" />
                <div className=\\"col\\">
                  <button className=\\"btn generalButton deleteButton\\" onClick={[Function: onClick]}>
                    Delete product
                  </button>
                </div>
                <div className=\\"col\\">
                  <button className=\\"btn generalButton allowButton\\" onClick={[Function: onClick]}>
                    Allow edit
                  </button>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>


<div className=\\"col d-md-flex colInContainer\\">
  <div className=\\"card mb-3 cardStyle\\">
    <div className=\\"row no-gutters justify-content-center\\">
      <div className=\\"col-md-6\\">
        <img src={{...}} className=\\"card-img insideCard\\" alt=\\"T-Shirt\\" />
      </div>
      <div className=\\"col-md-8\\">
        <div className=\\"card-body\\">
          <div className=\\"alert alert-info\\" role=\\"alert\\">
            Clothes already sold
          </div>
          <h5 className=\\"card-title\\">
            Main information
          </h5>
          <ul className=\\"list-group\\">
            <form>
              <div className=\\"form-group\\" />
            </form>
            <li className=\\"list-group-item\\">
              <div className=\\"row justify-content-center\\">
                <div className=\\"col\\" />
                <div className=\\"col\\">
                  <button className=\\"btn generalButton deleteButton\\" onClick={[Function: onClick]}>
                    Delete product
                  </button>
                </div>
                <div className=\\"col\\">
                  <button className=\\"btn generalButton allowButton\\" onClick={[Function: onClick]}>
                    Allow edit
                  </button>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
